Support ControlNet

ryu38 opened the pull request:
I added ControlNet feature in model conversion and inference.

**New Files**

controlnet.py
 - ControlNet Model on PyTorch (Reference: https://github.com/huggingface/diffusers/pull/2407)
 - To optimize ControlNet on CoreML, there are some changes from the original model, similar to UNet..

ControlNet.swift
- This is used in image generation with Swift.

**Main Changes**

torch2coreml.py
- two new options added
  - `--convert-contronet`
    - Unlike other --convert-* options, it requires controlnet models name after the option.
    - To convert multiple models, provide their names separated by spaces.
    - Example: `--convert-contronet lllyasviel/sd-controlnet-mlsd lllyasviel/sd-controlnet-canny`
    - ControlNet model is saved as `ControlNet_lllyasviel_sd-controlnet-mlsd.mlpackage`
  - `--unet-support-controlnet`
    - This option enables UNet to receive ControlNet results as additional inputs.
    - The model is saved with a different name: `*_control-unet.mlpackage`

unet.py and UNet.swift
- Supports ControlNet

pipeline.py
- two new options added
  - `--controlnet`
    - Models provided with this option are used in image generation.
    - Enter the option in the same way as `--convert-contronet` option in torch2coreml.py
  - `--controlnet-inputs`
    - Image inputs corresponding to each ControlNet
    - Enter paths to the images in same order as `--controlnet`
- If ControlNet is enabled, pipeline uses "control-unet.py" instead of "unet.py"

StableDiffusionCLI
- two new options added. These are almost the same as ones in pipeline.py.
  - `--controlnet` (enter model file names in Resources/controlnet without extension)
  - `--controlnet-inputs`
- If ControlNet is enabled, pipeline uses "ControledUNet.mlmodelc" instead of "UNet.mlmodelc"

Do not erase the below when submitting your pull request:
#########

- [x] I agree to the terms outlined in CONTRIBUTING.md

pj4533 commented:
🎉 nice!

atiorh commented on python_coreml_stable_diffusion/torch2coreml.py:
```diff
@@ -194,6 +208,9 @@ def bundle_resources_for_swift_cli(args):
                                      ("unet", "Unet"),
                                      ("unet_chunk1", "UnetChunk1"),
                                      ("unet_chunk2", "UnetChunk2"),
+                                     ("control-unet", "ControledUnet"),
```
NIT: Could we please change this to`ControlledUnet`?

atiorh commented on python_coreml_stable_diffusion/torch2coreml.py:
```diff
@@ -598,23 +630,59 @@ def convert_unet(pipe, args):
+        load_state_dict_summary = reference_unet.load_state_dict(
+            pipe.unet.state_dict())
+
+        if args.unet_support_controlnet:
```
Is there any situation where `args.unet_support_controlnet` should be false when at least one `--convert-controlnet` argument is provided? Would it make sense to make `args.unet_support_controlnet` dependent on `--convert-controlnet` being provided?

ryu38 replied:
I assumed a situation like that we convert a model of Stable Diffusion to be compatible with ControlNet models which have already been converted along with another Stable Diffusion model. In that case, we wouldn't need to convert controlNet itself but making UNet support it would be required.

atiorh replied:
Understood! My understanding is that ControlNet models are tethered to specific Stable Diffusion versions. Could you let me know if that is an accurate assumption?

atiorh replied:
Either way, I think it should be fine if you would like to keep `args.unet_support_controlnet`.

ryu38 replied:
> ControlNet models are tethered to specific Stable Diffusion versions.

Yes, I think that's right. But controlnet, which was trained on a certain version of original stable diffusion, also looks working well when used for fine-tuned models derived from original one.
For example, https://huggingface.co/lllyasviel/sd-controlnet-mlsd was trained on SD 1.5 can be used for fine-tuned models derived from SD 1.5 as well.

atiorh commented on python_coreml_stable_diffusion/torch2coreml.py:
```diff
@@ -872,6 +940,169 @@ def cosine_distance(image_embeds, text_embeds):
+                    "encoder_hidden_states"].squeeze(2).transpose(1, 2)
+
+        # Import controlnet model and initialize reference controlnet
+        original_controlnet = ControlNetModel.from_pretrained(
```
I reviewed https://huggingface.co/lllyasviel/sd-controlnet-mlsd/blob/main/config.json to see if there was an indication of which `StableDiffusionModel` models a particular `ControlNetModel` model is compatible with but didn't find anything. Do you have a recommended way perform such a compatibility check?

ryu38 replied:
Hmm, comparing the shapes of ControlNet outputs with those of UNet additional inputs to ensure they are the same may work as a part of the compatibility check, but it does not completely assure compatibility.
I can check them while converting models in torch2coreml.py and warn if they are not compatible.

atiorh replied:
I meant more of a `config.json` level compatibility hint but it seems that this information does not currently exist. @pcuenca what do you think? This is a nice-to-have (not a merge blocker in my opinion)

pcuenca replied:
Thanks for the ping! I think it's a good idea to save the base model in the `config.json` of the trained ControlNet models, I'll create an issue in `diffusers` for discussion.

Regarding the initial collection of ControlNet models, as far as I know they were all trained using [Stable Diffusion v1.5](https://huggingface.co/runwayml/stable-diffusion-v1-5), so that could be a reasonable fallback default.

atiorh replied:
Thanks @pcuenca ! We can update this repo once that issue is resolved on the diffusers side. @ryu38 For the short term (merging this PR before diffusers issue is resolved), could you please add a `logger.warning` message mentioning the version compatibility risk and maybe call out v1.5 as the default for the ControlNet original paper models?

ryu38 replied:
I added the warning message as below in the new commit.
https://github.com/apple/ml-stable-diffusion/blob/f535554024706602f66c65a5e8b4b6205899afc3/python_coreml_stable_diffusion/torch2coreml.py#L956-L959

atiorh replied:
Great!

atiorh commented on swift/StableDiffusion/pipeline/CGImage+vImage.swift:
```diff
@@ -63,58 +63,56 @@ extension CGImage {
-            var destinationG = try vImage_Buffer(width: Int(width), height: Int(height), bitsPerPixel: 8 * UInt32(MemoryLayout<Float>.size))
-            var destinationB = try vImage_Buffer(width: Int(width), height: Int(height), bitsPerPixel: 8 * UInt32(MemoryLayout<Float>.size))
-            
-            var minFloat: [Float] = [-1.0, -1.0, -1.0, -1.0]
```
The diff in this file looks unexpectedly large, could you please verify that the only changes are related to minFloat and maxFloat vars?

alejandro-isaza commented on swift/StableDiffusion/pipeline/ControlNet.swift:
```diff
@@ -0,0 +1,136 @@
+            for n in 0..<results.count {
+                let result = results.features(at: n)
+                if currentOutputs.count < results.count {
+                    let initOutput = result.featureNames.reduce(into: [String: MLMultiArray]()) { output, k in
```
Let's use MLShapedArray instead of MLMultiArray

alejandro-isaza commented on swift/StableDiffusion/pipeline/ControlNet.swift:
```diff
@@ -0,0 +1,136 @@
+                let result = results.features(at: n)
+                if currentOutputs.count < results.count {
+                    let initOutput = result.featureNames.reduce(into: [String: MLMultiArray]()) { output, k in
+                        output[k] = MLMultiArray(
```
This would be a lot faster if we could pre-allocate the output with the expected size.

ryu38 replied:
Is this suggesting that we should pre-allocate MLShapedArray with a specific shape in output dictionary? If we do this before allocating model results, would we create an MLShapedArray filled with zero values?

alejandro-isaza replied:
Yes, create it with the right size and fill with zeros.

alejandro-isaza commented on swift/StableDiffusion/pipeline/StableDiffusionPipeline+Resources.swift:
```diff
@@ -56,15 +66,38 @@ public extension StableDiffusionPipeline {
+            let fileName = model + ".mlmodelc"
+            return urls.controlNetDirURL.appending(path: fileName)
+        }
+        if (!controlNetURLs.isEmpty) {
```
```suggestion
        if !controlNetURLs.isEmpty {
```

alejandro-isaza commented on swift/StableDiffusion/pipeline/StableDiffusionPipeline+Resources.swift:
```diff
@@ -56,15 +66,38 @@ public extension StableDiffusionPipeline {
+        let unetURL: URL, unetChunk1URL: URL, unetChunk2URL: URL
+        
+        // if ControlNet available, Unet supports additional inputs from ControlNet
+        if (controlNet == nil) {
```
```suggestion
        if controlNet == nil {
```

alejandro-isaza commented on swift/StableDiffusion/pipeline/Unet.swift:
```diff
@@ -79,19 +79,23 @@ public struct Unet: ResourceManaging {
                 "timestep" : MLMultiArray(t),
                 "encoder_hidden_states": MLMultiArray(hiddenStates)
             ]
+            additionalResiduals?[$0.offset].forEach { (k, v) in
```
```suggestion
            for (k, v) int additionalResiduals?[$0.offset] {
```

imk2o commented on swift/StableDiffusion/pipeline/StableDiffusionPipeline+Resources.swift:
```diff
@@ -29,6 +33,10 @@ public extension StableDiffusionPipeline {
             safetyCheckerURL = baseURL.appending(path: "SafetyChecker.mlmodelc")
             vocabURL = baseURL.appending(path: "vocab.json")
             mergesURL = baseURL.appending(path: "merges.txt")
+            controlNetDirURL = baseURL.appending(path: "Controlnet")
```
Since torch2coreml seems to export to the `controlnet` directory, it seems like a good idea to start with lower case here as well.

> Thanks for your great contribution!

ryu38 commented:
Thank you for your reviews! I'll check or fix them one by one. I'll also update README to include about the new args.

atiorh commented:
@ryu38 I see that you have pushed some commits addressing the feedback. Please let me know when you would like me to re-review :)

ryu38 commented:
@atiorh Thank you! Could you please review `ControlNet.swift` to see if it can run faster or if there is any room for performance improvement?

atiorh commented:
Update: I am running the final tests and I will merge this PR when they pass. The latest commit seems to have addressed all the feedback but I will do one more visual pass just in case

ryu38 commented:
@atiorh I apologize that I pushed new commit just before the branch merged. This commit addressed the remaining feedback and improved inference speed in ControlNet.

pj4533 commented:
Just wow! Well done all. Can't wait to dig into this! 🎉🎉🎉🎉

atiorh commented:
Just realized the extra commit, this is my bad too! I don't have concerns with the diff though. Thanks for the contribution @ryu38 !

TimYao18 commented:
Excuse me, 

I call the following command:
```bash
python -m python_coreml_stable_diffusion.torch2coreml \
    --convert-vae-decoder --convert-vae-encoder --convert-unet \
    --unet-support-controlnet --convert-text-encoder \
    --model-version runwayml/stable-diffusion-v1-5 \
    --bundle-resources-for-swift-cli \
    --quantize-nbits 6 \
    --attention-implementation SPLIT_EINSUM_V2 \
    -o ~/MochiDiffusion/models && \
    python -m python_coreml_stable_diffusion.torch2coreml \
    --convert-unet --unet-support-controlnet \
    --model-version runwayml/stable-diffusion-v1-5 \
    --bundle-resources-for-swift-cli \
    --quantize-nbits 6 \
    --attention-implementation SPLIT_EINSUM_V2 \
    -o ~/MochiDiffusion/models
```

but only these files are generated, no **Unet**:
ControlledUnet.mlmodelc
TextEncoder.mlmodelc
VAEDecoder.mlmodelc
VAEEncoder.mlmodelc
merges.txt
vocab.json

If I want to get runnable model supported controlNet, what commands should I run?

jrittvo commented:
The files you ended up with are a working model, when used along with a ControlNet model.  But they won't work without a ControlNet model.  That is, they won't work for regular inference, or for Image2Image.  

To also get the `Unet.mlmodelc` so that the base model will work with and without a ControlNet in the pipeline, remove `--unet-support-controlnet` from the second command (the one after the `&&`).  That pass will now add the `Unet.mlmodelc` to the files from the first pass.

The `--unet-support-controlnet` modifies the type of of Unet created by the `--convert-unet` argument.  With just `--convert-unet`you get a `Unet.mlmodelc`.  With `--convert-unet` and `unet-support-controlnet` together, you get a `ControlledUnet.mlmodelc`.  With just `unet-support-controlnet` you do not get any Unet.
